[SPARK-19968][SS] Use a cached instance of `KafkaProducer` instead of creating one every batch.

ScrapCodes opened the pull request:
## What changes were proposed in this pull request?

In summary, cost of recreating a KafkaProducer for writing every batch is high as it starts a lot threads and make connections and then closes them. A KafkaProducer instance is promised to be thread safe in Kafka docs. Reuse of KafkaProducer instance while writing via multiple threads is encouraged.

Furthermore, I have performance improvement of 10x in latency, with this patch.

### These are times that addBatch took in ms. Without applying this patch
![with-out_patch](https://cloud.githubusercontent.com/assets/992952/23994612/a9de4a42-0a6b-11e7-9d5b-7ae18775bee4.png)
### These are times that addBatch took in ms. After applying this patch
![with_patch](https://cloud.githubusercontent.com/assets/992952/23994616/ad8c11ec-0a6b-11e7-8634-2266ebb5033f.png)


## How was this patch tested?
Running distributed benchmarks comparing runs with this patch and without it.
Added relevant unit tests.

SparkQA commented:
**[Test build #74644 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/74644/testReport)** for PR 17308 at commit [`febf387`](https://github.com/apache/spark/commit/febf3874cf07bad04e574b571f1caa839c9c28b7).

ScrapCodes commented on external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaWriteTask.scala:
```diff
@@ -32,7 +31,7 @@ import org.apache.spark.sql.types.{BinaryType, StringType}
  */
 private[kafka010] class KafkaWriteTask(
-    producerConfiguration: ju.Map[String, Object],
+    producerConfiguration: ju.HashMap[String, Object],
```
Ideally this should not have been changed. But, they are changed to HashMap, to avoid converting or casting them later.

SparkQA commented:
**[Test build #74646 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/74646/testReport)** for PR 17308 at commit [`830d76d`](https://github.com/apache/spark/commit/830d76d3ce19aa7fdcafed2a3c67338232d9dbeb).

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/74644/
Test FAILed.

SparkQA commented:
**[Test build #74646 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/74646/testReport)** for PR 17308 at commit [`830d76d`](https://github.com/apache/spark/commit/830d76d3ce19aa7fdcafed2a3c67338232d9dbeb).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/74646/
Test FAILed.

SparkQA commented:
**[Test build #74658 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/74658/testReport)** for PR 17308 at commit [`0b5e2e8`](https://github.com/apache/spark/commit/0b5e2e82d8f190bbf115234511f9c8e24f8e1995).

SparkQA commented:
**[Test build #74658 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/74658/testReport)** for PR 17308 at commit [`0b5e2e8`](https://github.com/apache/spark/commit/0b5e2e82d8f190bbf115234511f9c8e24f8e1995).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class CachedKafkaProducerSuite extends SharedSQLContext with PrivateMethodTester `

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/74658/
Test PASSed.

ScrapCodes commented:
Please take a look, @tcondie @zsxwing !

ScrapCodes commented:
@tcondie and @zsxwing any comments on this patch. I would be happy, if this bug is fixed before 2.2 is released.

ScrapCodes commented:
@tdas ping !

ScrapCodes commented:
I can further confirm, that in logs,  a kafkaproducer instance is created almost every instant.

BenFradet commented on external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/CachedKafkaProducer.scala:
```diff
@@ -0,0 +1,70 @@
+    producerConfiguration: ju.HashMap[String, Object]): KafkaProducer[Array[Byte], Array[Byte]] = {
+    val kafkaProducer: KafkaProducer[Array[Byte], Array[Byte]] =
+      new KafkaProducer[Array[Byte], Array[Byte]](producerConfiguration)
+    cacheMap.put(producerConfiguration.hashCode(), kafkaProducer)
```
Just curious if this a good idea to key the producer map by the hash code of a map for which the values are `Object`s?

ScrapCodes replied:
It is not a good idea to do like that. 

I had like my understanding to be corrected, as much as I understood. Since in this particular case Spark does not let user specify a key or value serializer/deserializer. So `Object` can be either a String, int or Long and for these hashcode would work correctly. I am also contemplating a better way to do it, now.

BenFradet replied:
True, my bad I thought `KafkaSink` was a public API.

marmbrus replied:
Yeah, I don't think this is a good key for the hashmap.  There could be collisions.  We should either assign a unique ID to the sink and thread that through, or come up with some way to canoncicalize the set of parameters that create the sink.  The latter might better since you could maybe reuse the same producer for more than one query.

koeninger commented:
Just to throw in my two cents, a change like this is definitely needed, as is made clear by the second sentence of the docs

http://kafka.apache.org/0102/javadoc/index.html?org/apache/kafka/clients/producer/KafkaProducer.html

"The producer is thread safe and sharing a single producer instance across threads will generally be faster than having multiple instances."

koeninger commented:
@marmbrus @zsxwing @tdas This needs attention from someone

brkyvz commented:
Taking a look

marmbrus commented on external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaSink.scala:
```diff
@@ -30,14 +30,19 @@ private[kafka010] class KafkaSink(
   }
+
+  override def stop(): Unit = {
+    CachedKafkaProducer.close(kafkaParams)
```
This is only closing the producer on the driver, right?  Do we even create on there?

ScrapCodes replied:
That's correct, I have understood, close requires a bit of rethinking, I am unable to see a straight forward way of doing it. If you agree, close related implementation can be taken out from this PR and be taken up in a separate JIRA and PR ?

marmbrus commented on external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/CachedKafkaProducer.scala:
```diff
@@ -0,0 +1,70 @@
+   * one instance per specified kafkaParams.
+   */
+  def getOrCreate(kafkaParams: ju.HashMap[String, Object])
+  : KafkaProducer[Array[Byte], Array[Byte]] = synchronized {
```
nit: Typically we wrap the arguments rather than the return type

brkyvz commented on external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/CachedKafkaProducer.scala:
```diff
@@ -0,0 +1,70 @@
+
+private[kafka010] object CachedKafkaProducer extends Logging {
+
+  private val cacheMap = new mutable.HashMap[Int, KafkaProducer[Array[Byte], Array[Byte]]]()
```
uber nit: maybe we can define a type alias
```scala
type Producer = KafkaProducer[Array[Byte], Array[Byte]]
```
so that we don't have to write that whole thing over and over

marmbrus reviewed (changes requested):
I agree this is an important optimization we need to do.  I have some concerns about the life cycle of the producer as its implemented here.

brkyvz commented on external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/CachedKafkaProducer.scala:
```diff
@@ -0,0 +1,70 @@
+    // KafkaParams is a java.util.HashMap as it produces a new hashcode for
+    // for every change in the contents of the HashMap and ensures m1 == m2 implies
+    // m1.hashCode == m2.hashCode.
+    log.debug(s"Created a new instance of KafkaProducer for $kafkaParams.")
```
maybe put this in the `create` function instead

brkyvz commented on external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaWriteTask.scala:
```diff
@@ -88,7 +85,7 @@ private[kafka010] class KafkaWriteTask(
         throw new IllegalStateException(s"${KafkaWriter.TOPIC_ATTRIBUTE_NAME} " +
           s"attribute unsupported type $t. ${KafkaWriter.TOPIC_ATTRIBUTE_NAME} " +
-          s"must be a ${StringType}")
+          s"must be a $StringType")
```
don't need the `s` and `$` right?

ScrapCodes commented:
Hi @marmbrus and @brkyvz, Thanks a lot of taking a look.

@marmbrus You are right, we should have another way to canonicalize kafka params. I can only think of appending a unique id to kafka params and somehow ensuring a particular set of params get the same uid everytime.

SparkQA commented:
**[Test build #76868 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/76868/testReport)** for PR 17308 at commit [`932d563`](https://github.com/apache/spark/commit/932d5638feed4fd5c602da27129b5bc4b81ff897).

ScrapCodes committed:
[SPARK-19968][SS] Use a cached instance of KafkaProducer instead of creating one every batch.

SparkQA commented:
**[Test build #76868 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/76868/testReport)** for PR 17308 at commit [`932d563`](https://github.com/apache/spark/commit/932d5638feed4fd5c602da27129b5bc4b81ff897).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/76868/
Test PASSed.

SparkQA commented:
**[Test build #76869 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/76869/testReport)** for PR 17308 at commit [`2a15afe`](https://github.com/apache/spark/commit/2a15afe326ba7660ff991811fa200c796113b146).

SparkQA commented:
**[Test build #76869 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/76869/testReport)** for PR 17308 at commit [`2a15afe`](https://github.com/apache/spark/commit/2a15afe326ba7660ff991811fa200c796113b146).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/76869/
Test PASSed.

SparkQA commented:
**[Test build #76872 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/76872/testReport)** for PR 17308 at commit [`ea9592a`](https://github.com/apache/spark/commit/ea9592adacfa7b923d1df93d3b40f0ec6a3bc548).

SparkQA commented:
**[Test build #76872 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/76872/testReport)** for PR 17308 at commit [`ea9592a`](https://github.com/apache/spark/commit/ea9592adacfa7b923d1df93d3b40f0ec6a3bc548).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/76872/
Test FAILed.

SparkQA commented:
**[Test build #76890 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/76890/testReport)** for PR 17308 at commit [`8224596`](https://github.com/apache/spark/commit/82245967eb618e3bdd865c200df7aaca2e41a7a6).

SparkQA commented:
**[Test build #76890 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/76890/testReport)** for PR 17308 at commit [`8224596`](https://github.com/apache/spark/commit/82245967eb618e3bdd865c200df7aaca2e41a7a6).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/76890/
Test PASSed.

ScrapCodes commented:
SPARK-20737 is created to look into cleanup mechanism in a separate JIRA.

ScrapCodes committed:
Self review and code style improvements.

SparkQA commented:
**[Test build #76932 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/76932/testReport)** for PR 17308 at commit [`e07e77e`](https://github.com/apache/spark/commit/e07e77e78c568a7cb5e68eb25e71e330a9f61962).

SparkQA commented:
**[Test build #76932 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/76932/testReport)** for PR 17308 at commit [`e07e77e`](https://github.com/apache/spark/commit/e07e77e78c568a7cb5e68eb25e71e330a9f61962).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/76932/
Test PASSed.

ScrapCodes committed:
[SPARK-20737] Mechanism for cleanup hooks, for structured-streaming sinks on executor shutdown.

Add a standard way of cleanup during shutdown of executors for structured streaming sinks in general and KafkaSink in particular.

ScrapCodes commented on external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaWriter.scala:
```diff
@@ -94,4 +94,10 @@ private[kafka010] object KafkaWriter extends Logging {
+  def close(sc: SparkContext, kafkaParams: ju.Map[String, Object]): Unit = {
+    sc.parallelize(1 to 10000).foreachPartition { iter =>
+      CachedKafkaProducer.close(kafkaParams)
+    }
```
This would cause  `CachedKafkaProducer.close` to be executed on each executor. I am thinking of a better way here.

brkyvz replied:
AFAIK the KafkaSource also faces the same issue of not being able to close consumers. Can we use a guava cache with a (configurable) timeout? I guess that's the safest way to make sure that they'll eventually get closed.

ScrapCodes replied:
Using guave cache, we can close if not used for a certain time. Shall we ignore closing them during a shutdown ? 
In the particular case of kafka producer, I do not see a direct problem with that. Since we do a producer.flush() on each batch. I was just wondering, with streaming sinks general - what should be our strategy ?

ScrapCodes committed:
Synchronized access to local hash map.

ScrapCodes commented:
@marmbrus and @brkyvz Please take another look, and let me know how it can be improved further. Thank you for the help so far.

SparkQA commented:
**[Test build #77021 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/77021/testReport)** for PR 17308 at commit [`3ec9981`](https://github.com/apache/spark/commit/3ec9981d4148b58df1f0e7c029762d5a6d3dd446).

brkyvz commented on external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/CachedKafkaProducer.scala:
```diff
@@ -0,0 +1,148 @@
+  private val cacheMap = new mutable.HashMap[String, Producer]()
+
+  private def createKafkaProducer(
+    producerConfiguration: ju.Map[String, Object]): Producer = {
```
nit: indent 2 more spaces
```scala
private def createKafkaProducer(
    producerConfiguration: ju.Map[String, Object]): Producer = {
  ...
}
```

brkyvz commented on external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/CachedKafkaProducer.scala:
```diff
@@ -0,0 +1,148 @@
+   * one instance per specified kafkaParams.
+   */
+  private[kafka010] def getOrCreate(
+    kafkaParams: ju.Map[String, Object]): Producer = synchronized {
```
ditto

brkyvz commented on external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/CachedKafkaProducer.scala:
```diff
@@ -0,0 +1,148 @@
+  }
+
+  private[kafka010] def isCanonicalized(kafkaParams: ju.Map[String, Object]): Boolean = {
+    if (kafkaParams.get(sparkKafkaParamsUniqueId) != null) {
```
nit: you don't need the `if-else`

brkyvz commented on external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/CachedKafkaProducer.scala:
```diff
@@ -0,0 +1,148 @@
+  }
+
+  private[kafka010] def computeUniqueCanonicalForm(
+    kafkaParams: ju.Map[String, Object]): ju.Map[String, Object] = synchronized {
```
ditto on indent

brkyvz commented on external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/CachedKafkaProducer.scala:
```diff
@@ -0,0 +1,148 @@
+  private[kafka010] def computeUniqueCanonicalForm(
+    kafkaParams: ju.Map[String, Object]): ju.Map[String, Object] = synchronized {
+    if (isCanonicalized(kafkaParams)) {
+      logWarning(s"A unique id,$sparkKafkaParamsUniqueId ->" +
```
nit: space after `,`

brkyvz commented on external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/CachedKafkaProducer.scala:
```diff
@@ -0,0 +1,148 @@
+    if (isCanonicalized(kafkaParams)) {
+      logWarning(s"A unique id,$sparkKafkaParamsUniqueId ->" +
+        s" ${kafkaParams.get(sparkKafkaParamsUniqueId)}" +
+        s" already exists in kafka params, returning Kafka Params:$kafkaParams as is.")
```
space after `:`

brkyvz commented on external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/CachedKafkaProducer.scala:
```diff
@@ -0,0 +1,148 @@
+  private[kafka010] def clear(): Unit = {
+    registryMap.clear()
+  }
+
```
nit: no need for newline

brkyvz commented on external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaSource.scala:
```diff
@@ -70,13 +70,13 @@ import org.apache.spark.unsafe.types.UTF8String
-                                     metadataPath: String,
-                                     startingOffsets: KafkaOffsetRangeLimit,
-                                     failOnDataLoss: Boolean)
+  sqlContext: SQLContext,
```
4 spaces please

brkyvz reviewed (commented):
Regarding closing the producers, the KafkaSource faces the same issue. We don't close consumers there, but I think we can use a Guava Cache with a timeout to close producers. What do you think?
Also Left some comments related to style.

SparkQA commented:
**[Test build #77021 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/77021/testReport)** for PR 17308 at commit [`3ec9981`](https://github.com/apache/spark/commit/3ec9981d4148b58df1f0e7c029762d5a6d3dd446).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/77021/
Test PASSed.

viirya commented on sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamExecution.scala:
```diff
@@ -683,6 +685,17 @@ class StreamExecution(
 
+  /** Stops streaming sink safely. */
+  private def stopSink(): Unit = {
+      try {
```
nit: the ident is not the typical one.

viirya commented on external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/CachedKafkaProducer.scala:
```diff
@@ -0,0 +1,148 @@
+
+  private def createKafkaProducer(
+    producerConfiguration: ju.Map[String, Object]): Producer = {
+    val uid = producerConfiguration.get(CanonicalizeKafkaParams.sparkKafkaParamsUniqueId)
```
Shall we always use `getUniqueId` to get uid?

viirya commented on external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaSink.scala:
```diff
@@ -22,6 +22,7 @@ import java.{util => ju}
 import org.apache.spark.internal.Logging
 import org.apache.spark.sql.{DataFrame, SQLContext}
 import org.apache.spark.sql.execution.streaming.Sink
+import org.apache.spark.util.Utils
```
Do we need to import this?

viirya commented on external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/CanonicalizeKafkaParamsSuite.scala:
```diff
@@ -0,0 +1,61 @@
+    val kp2 = CanonicalizeKafkaParams.computeUniqueCanonicalForm(kafkaParams)
+    val kafkaParams2 = new ju.HashMap[String, Object](kafkaParams)
+    kafkaParams2.put("acks", "1")
+    val kp3 = CanonicalizeKafkaParams.computeUniqueCanonicalForm(kafkaParams2)
```
CanonicalKP2?

viirya commented on external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/CanonicalizeKafkaParamsSuite.scala:
```diff
@@ -0,0 +1,61 @@
+    kafkaParams.put("bootstrap.servers", "127.0.0.1:9022")
+    kafkaParams.put("key.serializer", classOf[ByteArraySerializer].getName)
+    kafkaParams.put("value.serializer", classOf[ByteArraySerializer].getName)
+    val kp2 = CanonicalizeKafkaParams.computeUniqueCanonicalForm(kafkaParams)
```
CanonicalKP?

ScrapCodes committed:
Style feedback

ScrapCodes commented:
@brkyvz I am toying with Guava cache, the only problem is, it does not have an in-built mechanism to do any cleanup on shutdown.

ScrapCodes commented:
@viirya Thank you for taking a look.

ScrapCodes committed:
Implementation using guava cache.

SparkQA commented:
**[Test build #77183 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/77183/testReport)** for PR 17308 at commit [`e5cd1e6`](https://github.com/apache/spark/commit/e5cd1e607748163cec4a2131ae23224e60317a9d).

SparkQA commented:
**[Test build #77183 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/77183/testReport)** for PR 17308 at commit [`e5cd1e6`](https://github.com/apache/spark/commit/e5cd1e607748163cec4a2131ae23224e60317a9d).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/77183/
Test PASSed.

brkyvz commented:
@ScrapCodes It should have a `clear` method for testing. We won't be able to guarantee shutdown being called on the executors when the `stop` method for the Sink is called anyway, that's why I suggested the timeout solution. If you really want, I think you can also use the `ShutdownHookManager` and call clean on the cache as well.

brkyvz commented on external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/CachedKafkaProducer.scala:
```diff
@@ -18,34 +18,58 @@
 
-  @GuardedBy("this")
-  private val cacheMap = new mutable.HashMap[String, Producer]()
+  private val cacheExpireTimeout: Long = System.getProperty("spark.kafka.guava.cache.timeout",
```
I wonder if this should be an option of the sink? I may be running multiple streams in a single cluster with different trigger intervals which would require different configs.

brkyvz replied:
Scratch that. We will only have a single cache...

brkyvz commented on external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/CachedKafkaProducer.scala:
```diff
@@ -18,34 +18,58 @@
+    override def onRemoval(notification: RemovalNotification[String, Option[Producer]]): Unit = {
+      val uid: String = notification.getKey
+      val producer = notification.getValue
+      log.debug(s"Evicting kafka producer $producer uid:$uid, due to ${notification.getCause}")
```
space after `:`

brkyvz commented on external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/CachedKafkaProducer.scala:
```diff
@@ -18,34 +18,58 @@
+  private val cacheExpireTimeout: Long = System.getProperty("spark.kafka.guava.cache.timeout",
+    "10").toLong
+
+  private val noneReturningLoader = new CacheLoader[String, Option[Producer]] {
```
The builder shouldn't **require** a loader.
https://google.github.io/guava/releases/16.0/api/docs/com/google/common/cache/CacheBuilder.html#build()
Not sure we use the same guava version though

ScrapCodes committed:
Added shutdown hook as stop at driver is not very reliable.

SparkQA commented:
**[Test build #77225 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/77225/testReport)** for PR 17308 at commit [`ef2d6cd`](https://github.com/apache/spark/commit/ef2d6cd4275d93518ec27d4b08916575a3e597d7).

ScrapCodes committed:
Corrected some style issues.

ScrapCodes committed:
removed a stray change.

SparkQA commented:
**[Test build #77227 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/77227/testReport)** for PR 17308 at commit [`15dfc80`](https://github.com/apache/spark/commit/15dfc80a8a35208f5f9df150de7c4bd9a015e2d8).

SparkQA commented:
**[Test build #77227 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/77227/testReport)** for PR 17308 at commit [`15dfc80`](https://github.com/apache/spark/commit/15dfc80a8a35208f5f9df150de7c4bd9a015e2d8).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/77227/
Test PASSed.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/77225/
Test FAILed.

ScrapCodes commented:
Jenkins, retest this please.

SparkQA commented:
**[Test build #77233 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/77233/testReport)** for PR 17308 at commit [`15dfc80`](https://github.com/apache/spark/commit/15dfc80a8a35208f5f9df150de7c4bd9a015e2d8).

SparkQA commented:
**[Test build #77233 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/77233/testReport)** for PR 17308 at commit [`15dfc80`](https://github.com/apache/spark/commit/15dfc80a8a35208f5f9df150de7c4bd9a015e2d8).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/77233/
Test PASSed.

brkyvz commented on external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/CachedKafkaProducer.scala:
```diff
@@ -0,0 +1,174 @@
+  private type Producer = KafkaProducer[Array[Byte], Array[Byte]]
+
+  private val cacheExpireTimeout: Long =
+    System.getProperty("spark.kafka.guava.cache.timeout", "10").toLong
```
probably useful to mention this is in minutes

brkyvz commented on external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaWriter.scala:
```diff
@@ -19,9 +19,9 @@ package org.apache.spark.sql.kafka010
 
 import java.{util => ju}
 
+import org.apache.spark.SparkContext
```
unnecessary import?

ScrapCodes committed:
Code review, feedback.

SparkQA commented:
**[Test build #77293 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/77293/testReport)** for PR 17308 at commit [`039d063`](https://github.com/apache/spark/commit/039d063af502586109afb0ecd135390c4b7d2050).

SparkQA commented:
**[Test build #77293 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/77293/testReport)** for PR 17308 at commit [`039d063`](https://github.com/apache/spark/commit/039d063af502586109afb0ecd135390c4b7d2050).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/77293/
Test PASSed.

viirya commented on external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/CachedKafkaProducer.scala:
```diff
@@ -0,0 +1,174 @@
+  }
+
+  private val guavaCache: Cache[String, Producer] = CacheBuilder.newBuilder()
+    .recordStats()
```
Do we use the stats?

viirya commented on external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/CanonicalizeKafkaParamsSuite.scala:
```diff
@@ -0,0 +1,61 @@
+import java.{util => ju}
+
+import org.apache.kafka.common.serialization.ByteArraySerializer
+import org.scalatest.PrivateMethodTester
```
Do we use this import?

ScrapCodes replied:
Ahh, oversight. Thanks !

brkyvz commented on external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/CachedKafkaProducer.scala:
```diff
@@ -36,7 +36,7 @@ private[kafka010] object CachedKafkaProducer extends Logging {
 
   private val cacheExpireTimeout: Long =
-    System.getProperty("spark.kafka.guava.cache.timeout", "10").toLong
+    System.getProperty("spark.kafka.guava.cache.timeout.minutes", "10").toLong
```
don't we need to get this from `SparkEnv` by the way? I don't know if the properties get populated properly.
Also, adding `minutes` to the conf makes it kinda long right? I think we can also replace `guava` with `producer`. 
I think it may also be better to use [this](https://github.com/apache/spark/blob/master/common/network-common/src/main/java/org/apache/spark/network/util/JavaUtils.java#L213) so that we get rid of `minutes` and users can actually provide arbitrary durations (hours if they want). I think that's what we generally use for `duration` type confs.

ScrapCodes replied:
Thanks, you are right !

ScrapCodes committed:
Using spark conf instead of System.getProperties.

SparkQA commented:
**[Test build #77297 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/77297/testReport)** for PR 17308 at commit [`1c9f892`](https://github.com/apache/spark/commit/1c9f89226ef37d338032fbf8b4b8a1c527cc5efd).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/77297/
Test PASSed.

brkyvz commented:
LGTM! @marmbrus @viirya do you have any more feedback?

zsxwing commented on external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/CachedKafkaProducer.scala:
```diff
@@ -0,0 +1,174 @@
+    .build[String, Producer]()
+
+  ShutdownHookManager.addShutdownHook { () =>
+    clear()
```
Do we really need to stop producers in a shutdown hook? I'm asking because stopping a producer is a blocking call and may prevent other shutdown hooks to run.

marmbrus replied:
+1, this seems complicated.  What exactly does shutdown do?  Is it just cleaning up thread pools?

zsxwing replied:
I think it will close connections as well. That's really not necessary since the process is being shut down.

marmbrus commented on external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/CachedKafkaProducer.scala:
```diff
@@ -0,0 +1,174 @@
+  }
+
+  private def createKafkaProducer(
+    producerConfiguration: ju.Map[String, Object]): Producer = {
```
nit: indent 4 here

marmbrus commented on external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/CachedKafkaProducer.scala:
```diff
@@ -0,0 +1,174 @@
+ * not exist. This is done to ensure, we have only one set of kafka parameters associated with a
+ * unique ID.
+ */
+private[kafka010] object CanonicalizeKafkaParams extends Logging {
```
This seems kind of complicated also.  Since we know these are always coming from `Data[Stream/Frame]Writer` and that will always give you `Map[String, String]` (and we expect the number of options here to be small).  Could we just make the key for the cache a sorted `Seq[(String, String)]` rather than invent another GUID?

marmbrus reviewed (commented):
Thanks for working on this.  Its a huge latency improvement and I'll be using it next week at Spark Summit.

I just have one suggestion about the cache implementation.

ScrapCodes committed:
Michael's feedback!

SparkQA commented:
**[Test build #77357 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/77357/testReport)** for PR 17308 at commit [`588fa03`](https://github.com/apache/spark/commit/588fa03d299328405121534b114716cd00bc3a48).

SparkQA commented:
**[Test build #77357 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/77357/testReport)** for PR 17308 at commit [`588fa03`](https://github.com/apache/spark/commit/588fa03d299328405121534b114716cd00bc3a48).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/77357/
Test FAILed.

ScrapCodes commented:
Jenkins, retest this please !

SparkQA commented:
**[Test build #77358 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/77358/testReport)** for PR 17308 at commit [`588fa03`](https://github.com/apache/spark/commit/588fa03d299328405121534b114716cd00bc3a48).

SparkQA commented:
**[Test build #77358 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/77358/testReport)** for PR 17308 at commit [`588fa03`](https://github.com/apache/spark/commit/588fa03d299328405121534b114716cd00bc3a48).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/77358/
Test FAILed.

ScrapCodes commented:
Build is failing due to "Our attempt to download sbt locally to build/sbt-launch-0.13.13.jar failed. Please install sbt manually from http://www.scala-sbt.org/"

ScrapCodes commented:
Jenkins, retest this please !

SparkQA commented:
**[Test build #77410 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/77410/testReport)** for PR 17308 at commit [`588fa03`](https://github.com/apache/spark/commit/588fa03d299328405121534b114716cd00bc3a48).

SparkQA commented:
**[Test build #77410 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/77410/testReport)** for PR 17308 at commit [`588fa03`](https://github.com/apache/spark/commit/588fa03d299328405121534b114716cd00bc3a48).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/77410/
Test PASSed.

ScrapCodes commented:
@marmbrus Thank you for taking a look again. Surely, shut down hook is not ideal for closing kafka producers. In fact, for the case of kafka sink, it might be correct to skip cleanup step. I have tried to address your comments.

viirya commented on external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/CachedKafkaProducer.scala:
```diff
@@ -0,0 +1,101 @@
+    Option(guavaCache.getIfPresent(paramsSeq)).getOrElse(createKafkaProducer(kafkaParams))
+  }
+
+  def paramsToSeq(kafkaParams: ju.Map[String, Object]): Seq[(String, Object)] = {
```
nit: seems can be `private`?

viirya commented on external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/CachedKafkaProducerSuite.scala:
```diff
@@ -0,0 +1,75 @@
+
+    CachedKafkaProducer.close(kafkaParams)
+    val map2 = CachedKafkaProducer.invokePrivate(cacheMap())
+    assert(map2.size == 1)
```
We just know there is one KP by this assert. Seems we should also verify if we close the correct KP?

viirya commented on external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/CachedKafkaProducer.scala:
```diff
@@ -0,0 +1,101 @@
+
+  def paramsToSeq(kafkaParams: ju.Map[String, Object]): Seq[(String, Object)] = {
+    val paramsSeq: Seq[(String, Object)] =
+      kafkaParams.asScala.toSeq.sortBy(x => (x._1, x._2.toString))
```
As it is a map, seems we can just sort by x._1?

viirya commented:
LGTM and few minor comments.

zsxwing commented on external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaWriteTask.scala:
```diff
@@ -68,11 +67,10 @@ private[kafka010] class KafkaWriteTask(
-      checkForErrors
-      producer.close()
-      checkForErrors
-      producer = null
```
nit: please keep `producer = null` for double-close

zsxwing commented on external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/CachedKafkaProducer.scala:
```diff
@@ -0,0 +1,101 @@
+   */
+  private[kafka010] def getOrCreate(kafkaParams: ju.Map[String, Object]): Producer = synchronized {
+    val paramsSeq: Seq[(String, Object)] = paramsToSeq(kafkaParams)
+    Option(guavaCache.getIfPresent(paramsSeq)).getOrElse(createKafkaProducer(kafkaParams))
```
Remove `synchronized` and also throw inner exception instead after changing to use `LoadingCache`, such as
```
  private[kafka010] def getOrCreate(kafkaParams: ju.Map[String, Object]): Producer = {
    val paramsSeq: Seq[(String, Object)] = paramsToSeq(kafkaParams)
    try {
      guavaCache.get(paramsSeq)
    } catch {
      case e @ (_: ExecutionException | _: UncheckedExecutionException | _: ExecutionError)
        if e.getCause != null =>
          throw e.getCause
    }
  }
```

zsxwing commented on external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/CachedKafkaProducer.scala:
```diff
@@ -0,0 +1,101 @@
+  private lazy val guavaCache: Cache[Seq[(String, Object)], Producer] = CacheBuilder.newBuilder()
+    .expireAfterAccess(cacheExpireTimeout, TimeUnit.MILLISECONDS)
+    .removalListener(removalListener)
+    .build[Seq[(String, Object)], Producer]()
```
nit: Use `build(CacheLoader<? super K1, V1> loader)` to use LoadingCache, then `getOrCreate` will be very simple.

zsxwing reviewed (changes requested):
I suggest using LoadingCache to simplify the codes. Otherwise, looks good.

ScrapCodes committed:
Using Loading cache, and few style related fixes.

SparkQA commented:
**[Test build #77499 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/77499/testReport)** for PR 17308 at commit [`a10276a`](https://github.com/apache/spark/commit/a10276abc783ad3818821c31f38ce56acc3e9d77).

ScrapCodes commented:
Thanks @viirya and @zsxwing. I have tried to address you comments.

SparkQA commented:
**[Test build #77499 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/77499/testReport)** for PR 17308 at commit [`a10276a`](https://github.com/apache/spark/commit/a10276abc783ad3818821c31f38ce56acc3e9d77).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/77499/
Test PASSed.

zsxwing commented:
LGTM. Merging to master and 2.2. Thanks!
